fix: Update `avm/res/document-db/database-account`

seesharprun opened the pull request:
## Description

The current pipeline doesn't work for Azure Cosmos DB. Also the descriptions errantly refer to the service as "DocumentDB" or "CosmosDB". This is a maintenance update that will likely introduce a major 1.0 version. I will also attempt to fix the current PSRule warnings.

> - Closes #3195

- [x] Removes hard-coded location
- [x] Adds a way to set the minimal TLS version
- [x] Update defaults to remove PSRule warnings
- [x] Change `uniqueKeyPolicyKeys` parameter to be compliant with PSRule (cannot end with term "keys")
- [x] Test Azure deployment in my subscription

## Pipeline Reference

<!-- Insert your Pipeline Status Badge below -->

| Pipeline |
| -------- |
| [![avm.res.document-db.database-account ](https://github.com/seesharprun/bicep-registry-modules/actions/workflows/avm.res.document-db.database-account.yml/badge.svg?cosmos-db-document-db-fixes)](https://github.com/seesharprun/bicep-registry-modules/actions/workflows/avm.res.document-db.database-account.yml) |

## Type of Change

<!-- Use the checkboxes [x] on the options that are relevant. -->

- [x] Update to CI Environment or utilities (Non-module affecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [x] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [x] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [x] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

<!--  Please keep up to date with the contribution guide at https://aka.ms/avm/contribute/bicep -->

seesharprun commented:
Latest update, in my local pipelines, all static checks pass with no warnings. Unfortunately, this would cause us to rename one parameter since the original name violated a PSRule. I used this as justification to bump the version from `0.6` to `1.0`.

seesharprun commented:
Here's my latest local pipeline running in static tests only with no warnings or errors: https://github.com/seesharprun/bicep-registry-modules/actions/runs/11357050891

AlexanderSehr commented:
Hey @seesharprun, please note, we won't move from pre-release to release (1.0) just yet. I.e., 'breaking updates' should increase the minor version. There's a plan to go with the module to 1.0 based on a number of criteria after which we would follow a strict semver approach

eriqua commented on avm/res/document-db/database-account/version.json:
```diff
@@ -1,6 +1,6 @@
 {
     "$schema": "https://aka.ms/bicep-registry-module-version-file-schema#",
-    "version": "0.6",
+    "version": "1.0",
```
Hi @seesharprun please consider we are not yet releasing modules at version 1.0, so breaking changes are still to use minor updates together with backward compatible new features. Ref https://azure.github.io/Azure-Verified-Modules/specs/shared/#module-versioning-before-first-major-version-release-100

```suggestion
    "version": "0.7",
```

eriqua commented on avm/res/document-db/database-account/tests/e2e/analytical/main.test.bicep:
```diff
@@ -17,16 +17,15 @@ param serviceShort string = 'dddaanl'
-// Pipeline is selecting random regions which dont support all cosmos features and have constraints when creating new cosmos
-#disable-next-line no-hardcoded-location
-var enforcedLocation = 'eastasia'
+@description('Optional. The location to deploy resources to.')
```
Please consider that without an enforced location, the location will be randomly chosen among the intersection of allowed locations and this list of default ('eastus', 'uksouth', 'northeurope', 'eastasia').
You can check what this intersection would be from the logs. Before changing this we should make sure that any of the locations from that list would work

seesharprun replied:
@eriqua, right now `eastasia` and `northeurope` are throwing capacity errors for Azure Cosmos DB in my pipeline.

Would it be preferable to change the enforced location to something else? Can I choose a location outside of the allowed locations list?

CC: @bryansan-msft

seesharprun replied:
For now, I changed it to `eastus2` and will change it to something else if your team advises. This is the region I use in my own testing.

eriqua replied:
You can @seesharprun. Think of the allowed list more like a list common to the majority of services. You surely can enforce a different one on your tests, even a different one for each test if that is useful .
You can resolve this comment once read 🙂

eriqua commented on avm/res/document-db/database-account/tests/e2e/role/main.test.bicep:
```diff
@@ -19,23 +19,23 @@ param namePrefix string = '#_namePrefix_#'
 // Pipeline is selecting random regions which dont support all cosmos features and have constraints when creating new cosmos
 #disable-next-line no-hardcoded-location
-var enforcedLocation = 'eastus2'
+var resourceLocation = 'eastus2'
```
Is the intention to enforce a location just for this test? when using an enforced location please keep the `enforcedLocation` convention

seesharprun replied:
I'll make this change

seesharprun commented:
Updates today:

- [x] Instead or renaming `uniqueKeyPolicyKeys` to `uniqueKeyPolicyConfiguration`, I deprecated the first and added the second. This should allow us to increment a minor version instead of a major version. The first can be fully removed when you do go to **1.0**
- [x] Changed the enforced location to `eastus2` for now.

eriqua commented on avm/res/document-db/database-account/main.bicep:
```diff
@@ -885,11 +898,17 @@ type sqlDatabaseType = {
     throughput: int?
 
-    @description('Optional. The unique key policy configuration containing a list of unique keys that enforces uniqueness constraint on documents in the collection in the Azure Cosmos DB service.')
+    @description('Optional. [DEPRECATED] The unique key policy configuration containing a list of unique keys that enforces uniqueness constraint on documents in the collection in the Azure Cosmos DB service.')
```
What is the reason for deprecating in favor of a new name? Is it PSRule related by any chance?

seesharprun replied:
That's the only thing keeping me from all greens and no warnings (I have all greens on deployment and 1 warning now)

seesharprun replied:
BTW, I'm open to any other ideas. I just saw the requirement to have all green build with no warnings and this PSRule warning has been tricky to resolve.

eriqua replied:
Yes @seesharprun I agree and that's the same I'd have suggested. You followed the deprecation process in the best way and thanks a lot for that. I see that more applicable after v1.0.0 is released, but that's only my personal opinion. Other maintainers may have a different view. However for the time being, having the parameter name listed in the ps-rule.yml file is the best option. Thanks for the great contribution!

bryansan-msft replied:
To be completely honest I am more keen on just removing the deprecated parameter/code/test/whatever than accumulating it until we are allowed to bump the major version. We have in our AVM notes that the versions 0.*.* can contain breaking changes in any form.

![image](https://github.com/user-attachments/assets/5c06be6f-c68b-4044-81b9-d4fef69db133)

seesharprun replied:
How about this for a compromise:

We get the bug fix in soon, and do the breaking change in a different PR to rename the property.

Would that work?

eriqua commented on avm/utilities/pipelines/staticValidation/psrule/ps-rule.yaml:
```diff
@@ -70,6 +70,7 @@ configuration:
       "ssoClientSecretKeyVaultPath",
       "ssoSecretType",
       "tokenValidityLength",
+      "uniqueKeyPolicyKeys"
```
Awesome job again @seesharprun and great contribution. Since this update would trigger all pipelines, if you agree we'd like to have that modified before approving and merging this PR. Usually we test all modules whenever a change impacts shared files like this one, but in this case we think it would be an overkill.
Please bare with us managing that prior to approving this one 🙏

seesharprun replied:
I approve!

Thanks for the thorough feedback.

Once this is merged in, I can start on some contributions to add functionality. The goal here was to get to a working baseline.

Thanks again!

eriqua replied:
Keeping this open until the psrule fix gets merged

eriqua replied:
Done via #3587. @seesharprun please pull latest from upstream, you should have no changes needed on psrule.yml.
Please close this comment once done

seesharprun replied:
I just pulled latest

bryansan-msft commented:
@seesharprun just in case you haven't got the idea, but I have hardcoded the test region since the pipeline selects a random one and it might not have cosmos capacity resulting in the pipeline run failing. Also, sometimes the region that worked before suddenly stops working due to capacity constraints. I haven't been able to find a stable region.

seesharprun commented:
I switched to eastus2 which was more stable than eastasia in my pipelines.

You're right about capacity being a moving target.

eriqua reviewed (commented):
Thanks @seesharprun! 
@bryansan-msft to provide final review and approval 🙂

bryansan-msft commented:
> Thanks @seesharprun! @bryansan-msft to provide final review and approval 🙂

Waiting for the last 2 comments. Specially the one of the "defaults" test

seesharprun commented:
> > Thanks @seesharprun! @bryansan-msft to provide final review and approval 🙂
> 
> Waiting for the last 2 comments. Specially the one of the "defaults" test

@bryansan-msft - I can't find the 2 comments you need addressed. Do you have a link to them?

bryansan-msft commented:
> > > Thanks @seesharprun! @bryansan-msft to provide final review and approval 🙂
> > 
> > 
> > Waiting for the last 2 comments. Specially the one of the "defaults" test
> 
> @bryansan-msft - I can't find the 2 comments you need addressed. Do you have a link to them?

@seesharprun I didnt click the submit review button duh. sorry for the fuss

seesharprun commented on avm/res/document-db/database-account/tests/e2e/defaults/main.test.bicep:
```diff
@@ -44,6 +44,11 @@ module testDeployment '../../../main.bicep' = [
     params: {
       location: enforcedLocation
       name: '${namePrefix}${serviceShort}001'
+      sqlDatabases: [
```
This is required for PSRule. PSRule requires that any account has key-based auth and key-based metadata auth disabled.

In the `main.bicep` file, it only disables key-based auth or key-based metadata auth if the API is NoSQL or Table. This maps to the error you would get in ARM.

If I omit this, PSRule will have a warning that this account doesn't have key-based auth disabled. I could set a default in the main.bicep file, but I'm not sure if you would want that.

seesharprun replied:
The alternative is to update the sqlDatabases parameter in **main.bicep** to:

```bicep
@description('Optional. SQL Databases configurations.')
param sqlDatabases sqlDatabaseType[] = [
  {
    name: 'default-database'
  }
]
```

But that doesn't feel ideal either. I'm trying to get a clean build, but if you're comfortable allowing this PSRule to persist, we can just let the PSRule warning stay. The build will still be green just with 2 warnings on the `defaults` test.

seesharprun replied:
In the essence of speed, I will remove this parameter. We should revisit this in the future so we can clear all of the PSRule warnings. I'm happy to work on clearing the warnings prior to 1.0.

seesharprun commented on avm/res/document-db/database-account/main.bicep:
```diff
@@ -243,10 +255,11 @@ var databaseAccountProperties = union(
   ((!empty(sqlDatabases) || !empty(mongodbDatabases) || !empty(gremlinDatabases))
     ? {
-        // Common properties
+        // NoSQL, MongoDB RU, and Apache Gremlin properties
```
Updating this now.

bryansan-msft commented:
Merge whenever you see appropriate

seesharprun commented:
All of my pipelines are green. Merging now

AlexanderSehr commented:
Hey @bryansan-msft & @seesharprun,
please note that the [workflow](https://github.com/Azure/bicep-registry-modules/actions/runs/11390119682/job/31765081981) did fail in upstream main. I retriggered it already 4 times - but it's always the same test that wants to deploy to Central US that fails due to capacity reasons. 
While not the end of the world, this also means that the updates to the `main.bicep` were **not** published. To get them published we either keep on re-running the failed job - or create a new PR that also enforces a different location for the multi-region test & (important) introduces another change to the `main.bicep` file so that the `git diff` picks up on the module change during publishing.
